src/main.rs

```diff
@@ -1,23 +1,115 @@
-use llvm_ir::{Function, Module};
+use std::{collections::HashMap, option::Option::Some};
+
+use llvm_ir::{BasicBlock, Module, Operand, constant::Constant, instruction::Instruction, name::Name::{Name, Number}, operand::Operand::{LocalOperand, ConstantOperand, MetadataOperand}, terminator::Terminator};
+
+struct BFInterp {
+    mem: HashMap<String, u64>,
+}
+
+impl BFInterp {
+
+    fn new() -> BFInterp {
+        BFInterp { 
+            mem: HashMap::new(),
+        }
+    }
+
+    fn interpret(&mut self, module: Module) -> Result<(), &'static str> {  
+        let main_func = match module.get_func_by_name("main") {
+            Some(main) => main,
+            None => return Err("No main function"),
+        };
+
+        match self.it_bb(&main_func.name, &main_func.basic_blocks) {
+            Some(_ret) => return Ok(()),
+            None => return Ok(()),
+        }
+    }
+    
+    fn it_bb(&mut self, func_name: &String, bbs: &Vec<BasicBlock>) -> Option<u64> {
+        for bb in bbs.iter() {
+            for inst in bb.instrs.iter() {
+                match inst {
+                    Instruction::Alloca(_) => {},
+                    Instruction::Load(_) => {
+                        // mem.load(a, b)
+                        // todo!(),
+                    }
+                    Instruction::Store(store) => self.store_var(func_name, &store.address, &store.value),
+                    _ => todo!(),
+                }
+            }
+            match &bb.term {
+                Terminator::Ret(rec) => {
+                    match rec.return_operand.as_ref().unwrap() {
+                        LocalOperand { name, ty } => todo!(),
+                        ConstantOperand(con_ref) => {
+                            match con_ref.as_ref() {
+                                Constant::Int { bits, value } => {
+                                    return Some(*value);
+                                },
+                                _ => todo!(),
+                            }
+                        },
+                        MetadataOperand => todo!(),
+                    }
+                },
+                _ => todo!(),
+            }
+        } 
+        None
+    }
+
+    fn store_var(&mut self, func_name: &String, add: &Operand, val: &Operand) {
+        let add = self.add_str(func_name, add);
```

Comment:
I don't think we want to insert `String`s into the memory hashmap, especially since most locals will be integers. `Operand` is hashable, so you can just use it as a key as is.

Though I can see that you are also adding the function name below to the value, presumably to differentiate between locals of the same name in different functions. I suggest we don't worry about function calls for the moment (but when we do we will need to implement them via stack frames).

Reply:
Operand does not seem to be hashable.

Reply:
Sorry, my mistake. `Operand.Name` is hashable. I'm not entirely sure if we need to hold on to the type, but let's for now assume we don't.